design/decouple-ceph-version.md

```diff
@@ -0,0 +1,128 @@
+Rook 0.9 can also be tested to support upgrades to Nautilus since they may be released in the same time frame. Since the Ceph tools are forward compatible, the Mimic tools will be sufficient to support upgrading to Nautilus.
+If Nautilus is released after Rook 0.9, a patch release can be made to 0.9 so that Rook can officially support the upgrade at that point. The changes in the patch release should be minimal since upgrading to Nautilus could have been mostly planned for in 0.9.
+
+The operator will be made to understand differences in the Ceph versions that are necessary for orchestration. Some examples might include:
+- If running Luminous, start the Ceph dashboard on http. If running Mimic, a self-signed cert could be generated to start the dashboard with https.
+- If a new daemon is added in a future Ceph release, the operator would understand to deploy that daemon only if the Ceph version is at least that version.
+
+### Upgrades
+
+The flexibility during upgrades will now be improved since the upgrade of Rook will be independent from the upgrade to the Ceph version.
+- To upgrade Rook, update the version of the Rook operator container
+- To upgrade Ceph, make sure Rook is running the latest release, then update the `cephImage` in the cluster CRD
+
+Each time the operator starts, an idempotent orchestration is executed to ensure the cluster is in the desired state. As part of the orchestration, the version of the operator
+will be reviewed. If the version has changed, the operator will update each of the daemons in a predictable order such as: mon, mgr, osd, rgw, mds. If the Rook upgrade requires any special steps, they will be handled as each version upgrade requires.
+
+When the cluster CRD is updated with a new Ceph version, the same idempotent orchestration is executed to evaluated desired state that needs to be applied to the cluster.
+Over time as the operator becomes smarter and more versions are supported, the custom upgrade steps will be implemented as needed.
+
+Daemons will only be restarted when necessary for the upgrade. The Rook upgrade sometimes will not require a restart of the daemons,
+depending on if the pod spec changed. The Ceph upgrade will always require a restart of the daemons. In either case, a restart will be done in an orderly manner
+with one pod at a time along with health checks as the upgrade proceeds. The upgrade will be paused if the cluster becomes unhealthy.
+
+#### Admin control of upgrades
+
+To allow more control over the upgrade, we define `upgradePolicy` settings.
+- `newCephImage`: The version of the image to start applying to the daemons specified in the `upgrade` list. After the admin is satisfied with the upgrade, he should delete this property and the `upgrade` properties from the CRD to declare that Rook should fully be running on the new version.
```

Comment:
Shouldn't it be Rook that does this final modification when it has completed the upgrade?  Otherwise we're relying on the human to go verify the new version is running everywhere (which is work for them, and risks them getting it wrong)

Reply:
Good point, Rook should update this automatically if it sees that everything is updated. The user could also do it after the partial upgrade, which would allow Rook to finish the upgrade automatically.

Reply:
The flow around the `upgradePolicy` is suggesting that the desired state comes in phases that would give the admin control over upgrading the mon, mgr, osd, and other daemons in phases. After the orchestration would fully complete for each phase, the admin would verify that he is ready to continue upgrading the other daemons, then he would update the CRD with the next daemon in the desired state for the new version. When all daemons have been upgraded, there is no longer a need for the upgrade policy. The upgrade policy wouldn't need to be removed though since it's still the desired state.

This type of upgrade policy would certainly complicate the automation so I don't imagine it would be in the first release, but would provide for more advanced upgrade flexibility in the future.